File: src/core/router.ts
import path from 'node:path'
import {identifySite, parseUrlInput} from './domains.js'

export type TargetInspection =
  | {
      type: 'torrent'
      target: string
      name: string
    }
  | {
      type: 'direct'
      url: string
      filename: string
      size?: number
    }
  | {
      type: 'video'
      cleanUrl: string
      timeRange?: string
      timeLabel?: string
    }

const DIRECT_EXTENSIONS = new Set([
  'iso',
  'img',
  'zip',
  'tar',
  'gz',
  'bz2',
  'xz',
  '7z',
  'rar',
  'tgz',
  'bin',
  'pkg',
  'deb',
  'rpm',
  'appimage',
  'exe',
  'dmg',
  'pdf',
  'epub',
  'apk',
  'jar',
])

export function parseMagnetName(magnetUri: string): string {
  try {
    const u = new URL(magnetUri)
    const dn = u.searchParams.get('dn')
    if (dn) return dn.replace(/\+/g, ' ')
    const xt = u.searchParams.get('xt') ?? ''
    const hash = xt.split(':').pop() ?? 'torrent'
    return `Torrent (${hash.slice(0, 10)})`
  } catch {
    return 'BitTorrent Transfer'
  }
}

export function extractFilenameFromUrl(rawUrl: string): string {
  try {
    const u = new URL(rawUrl)
    const base = path.basename(u.pathname)
    return base || 'download'
  } catch {
    return 'download'
  }
}

export async function inspectTarget(
  rawInput: string,
  signal?: AbortSignal,
): Promise<TargetInspection> {
  const trimmed = rawInput.trim()

  // 1. BitTorrent & Magnet
  if (trimmed.startsWith('magnet:?')) {
    return {
      type: 'torrent',
      target: trimmed,
      name: parseMagnetName(trimmed),
    }
  }

  if (trimmed.endsWith('.torrent') || trimmed.includes('.torrent?')) {
    return {
      type: 'torrent',
      target: trimmed,
      name: path.basename(trimmed.split('?')[0]!, '.torrent'),
    }
  }

  // 2. Video / Streaming URL with optional trimming
  const {cleanUrl, timeRange, timeLabel} = parseUrlInput(trimmed)

  // If time range is specified or domain belongs to a known media site, route to video engine
  const site = identifySite(cleanUrl)
  if (timeRange || site.key !== 'generic') {
    return {
      type: 'video',
      cleanUrl,
      timeRange,
      timeLabel,
    }
  }

  // 3. Check for common direct download file extensions
  let pathname = ''
  try {
    pathname = new URL(cleanUrl).pathname.toLowerCase()
  } catch {
    return {type: 'video', cleanUrl}
  }

  const extMatch = /\.([a-z0-9]{2,8})$/i.exec(pathname)
  const ext = extMatch?.[1]?.toLowerCase()
  if (ext && DIRECT_EXTENSIONS.has(ext)) {
    return {
      type: 'direct',
      url: cleanUrl,
      filename: path.basename(pathname),
    }
  }

  // 4. Fast HEAD request for direct attachments / binary streams
  try {
    const headSignal = signal ? AbortSignal.any([signal, AbortSignal.timeout(1500)]) : AbortSignal.timeout(1500)
    const resp = await fetch(cleanUrl, {method: 'HEAD', signal: headSignal})
    if (resp.ok) {
      const disposition = resp.headers.get('content-disposition') ?? ''
      const contentType = (resp.headers.get('content-type') ?? '').toLowerCase()
      const lengthHeader = resp.headers.get('content-length')
      const size = lengthHeader ? Number.parseInt(lengthHeader, 10) : undefined

      const filenameMatch = /filename\*?=(?:UTF-8'')?"?([^";]+)"?/i.exec(disposition)
      const dispositionFilename = filenameMatch?.[1]

      const isAttachment = disposition.includes('attachment')
      const isBinary =
        contentType.includes('application/octet-stream') ||
        contentType.includes('application/x-') ||
        contentType.includes('application/zip') ||
        (!contentType.includes('text/html') && !contentType.includes('application/xhtml+xml'))

      if ((isAttachment || isBinary) && (dispositionFilename || (size && size > 1_000_000))) {
        return {
          type: 'direct',
          url: cleanUrl,
          filename: dispositionFilename || extractFilenameFromUrl(cleanUrl),
          size: Number.isFinite(size) ? size : undefined,
        }
      }
    }
  } catch {
    // network timeout or HEAD not allowed — fall through to video engine
  }

  return {
    type: 'video',
    cleanUrl,
    timeRange,
    timeLabel,
  }
}

File: internal/core/router.go
package core

import (
	"context"
	"fmt"
	"mime"
	"net/http"
	"net/url"
	"path"
	"strconv"
	"strings"
	"time"

	"github.com/KabosuNeko/anpan/internal/engine"
)

type TargetType string

const (
	TargetTorrent TargetType = "torrent"
	TargetDirect  TargetType = "direct"
	TargetArchive TargetType = "archive"
	TargetVideo   TargetType = "video"
)

type TargetInspection struct {
	Type TargetType

	// For torrent
	Target string
	Name   string

	// For direct
	URL      string
	Filename string
	Size     *int64

	// For archive
	ArchivePost *engine.ArchivePost

	// For video
	CleanURL  string
	TimeRange string
	TimeLabel string
}

var directExtensions = map[string]bool{
	"iso": true, "img": true, "zip": true, "tar": true, "gz": true,
	"bz2": true, "xz": true, "7z": true, "rar": true, "tgz": true,
	"bin": true, "pkg": true, "deb": true, "rpm": true, "appimage": true,
	"exe": true, "dmg": true, "pdf": true, "epub": true, "apk": true,
	"jar": true,
}

func ParseMagnetName(magnetURI string) string {
	u, err := url.Parse(magnetURI)
	if err != nil {
		return "BitTorrent Transfer"
	}
	dn := u.Query().Get("dn")
	if dn != "" {
		return strings.ReplaceAll(dn, "+", " ")
	}
	xt := u.Query().Get("xt")
	if xt != "" {
		parts := strings.Split(xt, ":")
		hash := parts[len(parts)-1]
		if len(hash) > 10 {
			hash = hash[:10]
		}
		return "Torrent (" + hash + ")"
	}
	return "BitTorrent Transfer"
}

func extractFilenameFromURL(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil {
		return "download"
	}
	base := path.Base(u.Path)
	if base == "" || base == "." || base == "/" {
		return "download"
	}
	return base
}

func InspectTarget(ctx context.Context, rawInput string) (*TargetInspection, error) {
	trimmed := strings.TrimSpace(rawInput)

	if strings.HasPrefix(trimmed, "magnet:?") {
		return &TargetInspection{
			Type:   TargetTorrent,
			Target: trimmed,
			Name:   ParseMagnetName(trimmed),
		}, nil
	}

	if strings.HasSuffix(trimmed, ".torrent") || strings.Contains(trimmed, ".torrent?") {
		parts := strings.Split(trimmed, "?")
		name := strings.TrimSuffix(path.Base(parts[0]), ".torrent")
		return &TargetInspection{
			Type:   TargetTorrent,
			Target: trimmed,
			Name:   name,
		}, nil
	}

	parsedInput := ParseURLInput(trimmed)
	cleanURL := parsedInput.CleanURL

	if engine.IsArchivePostURL(cleanURL) {
		archive, err := engine.ProbeArchivePost(ctx, cleanURL)
		if err != nil {
			return nil, fmt.Errorf("archive post could not be loaded: %w", err)
		}
		if archive != nil && len(archive.Files) > 0 {
			return &TargetInspection{
				Type:        TargetArchive,
				ArchivePost: archive,
			}, nil
		}
		return nil, fmt.Errorf("no downloadable files or attachments found in this post")
	}

	site := IdentifySite(cleanURL)
	if parsedInput.TimeRange != "" || site.Key != "generic" {
		return &TargetInspection{
			Type:      TargetVideo,
			CleanURL:  cleanURL,
			TimeRange: parsedInput.TimeRange,
			TimeLabel: parsedInput.TimeLabel,
		}, nil
	}

	u, err := url.Parse(cleanURL)
	if err != nil {
		return &TargetInspection{
			Type:     TargetVideo,
			CleanURL: cleanURL,
		}, nil
	}

	pathname := strings.ToLower(u.Path)
	ext := strings.TrimPrefix(path.Ext(pathname), ".")
	if directExtensions[ext] {
		return &TargetInspection{
			Type:     TargetDirect,
			URL:      cleanURL,
			Filename: path.Base(pathname),
		}, nil
	}

	headCtx, cancel := context.WithTimeout(ctx, 1500*time.Millisecond)
	defer cancel()

	req, err := http.NewRequestWithContext(headCtx, "HEAD", cleanURL, nil)
	if err == nil {
		client := &http.Client{Timeout: 1500 * time.Millisecond}
		resp, err := client.Do(req)
		if err == nil {
			defer resp.Body.Close()
			if resp.StatusCode >= 200 && resp.StatusCode < 300 {
				disposition := resp.Header.Get("Content-Disposition")
				contentType := strings.ToLower(resp.Header.Get("Content-Type"))
				lengthHeader := resp.Header.Get("Content-Length")
				var size *int64
				if lengthHeader != "" {
					if s, err := strconv.ParseInt(lengthHeader, 10, 64); err == nil {
						size = &s
					}
				}

				dispositionFilename := ""
				if disposition != "" {
					if _, params, err := mime.ParseMediaType(disposition); err == nil {
						dispositionFilename = params["filename"]
					}
				}

				isAttachment := strings.Contains(disposition, "attachment")
				isBinary := strings.Contains(contentType, "application/octet-stream") ||
					strings.Contains(contentType, "application/x-") ||
					strings.Contains(contentType, "application/zip") ||
					(!strings.Contains(contentType, "text/html") && !strings.Contains(contentType, "application/xhtml+xml"))

				if (isAttachment || isBinary) && (dispositionFilename != "" || (size != nil && *size > 1_000_000)) {
					filename := dispositionFilename
					if filename == "" {
						filename = extractFilenameFromURL(cleanURL)
					}
					return &TargetInspection{
						Type:     TargetDirect,
						URL:      cleanURL,
						Filename: filename,
						Size:     size,
					}, nil
				}
			}
		}
	}

	return &TargetInspection{
		Type:      TargetVideo,
		CleanURL:  cleanURL,
		TimeRange: parsedInput.TimeRange,
		TimeLabel: parsedInput.TimeLabel,
	}, nil
}
